From 85e89bbeff42c0195ec1f4cbbff063aeac145ad8 Mon Sep 17 00:00:00 2001
From: Takashi Yano <takashi.yano@nifty.ne.jp>
Date: Tue, 13 Feb 2024 11:36:05 +0900
Subject: [PATCH 58/N] Cygwin: pty: Fix potential handle leak regarding
 CallNamedPipe().

In pty master_thread, 6 handles are duplicated when CallNamedPipe()
requests that. Though some of them are not used so should be closed,
they were not. This causes handle leak potentially.

Backported-from: 73cd80c976 (Cygwin: pty: Fix potential handle leak regarding CallNamedPipe()., 2024-02-13)
Signed-off-by: Takashi Yano <takashi.yano@nifty.ne.jp>
Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>
---
 winsup/cygwin/fhandler/pty.cc | 20 ++++++++++++++++++--
 1 file changed, 18 insertions(+), 2 deletions(-)

diff --git a/winsup/cygwin/fhandler/pty.cc b/winsup/cygwin/fhandler/pty.cc
index 9d1a119..17b7b32 100644
--- a/winsup/cygwin/fhandler/pty.cc
+++ b/winsup/cygwin/fhandler/pty.cc
@@ -940,6 +940,8 @@ fhandler_pty_slave::open (int flags, mode_t)
 	  errmsg = "can't call master, %E";
 	  goto err;
 	}
+      CloseHandle (repl.to_slave_nat); /* not used. */
+      CloseHandle (repl.to_slave); /* not used. */
       from_master_nat_local = repl.from_master_nat;
       from_master_local = repl.from_master;
       to_master_nat_local = repl.to_master_nat;
@@ -1218,6 +1220,10 @@ fhandler_pty_slave::reset_switch_to_nat_pipe (void)
 		      if (!CallNamedPipe (pipe, &req, sizeof req,
 					  &repl, sizeof repl, &len, 500))
 			return; /* What can we do? */
+		      CloseHandle (repl.from_master); /* not used. */
+		      CloseHandle (repl.to_master); /* not used. */
+		      CloseHandle (repl.to_slave_nat); /* not used. */
+		      CloseHandle (repl.to_slave); /* not used. */
 		      CloseHandle (get_handle_nat ());
 		      set_handle_nat (repl.from_master_nat);
 		      CloseHandle (get_output_handle_nat ());
@@ -3932,10 +3938,20 @@ fhandler_pty_slave::transfer_input (tty::xfer_dir dir, HANDLE from, tty *ttyp,
       if (!CallNamedPipe (pipe, &req, sizeof req,
 			  &repl, sizeof repl, &len, 500))
 	return; /* What can we do? */
+      CloseHandle (repl.from_master_nat); /* not used. */
+      CloseHandle (repl.from_master); /* not used. */
+      CloseHandle (repl.to_master_nat); /* not used. */
+      CloseHandle (repl.to_master); /* not used. */
       if (dir == tty::to_nat)
-	to = repl.to_slave_nat;
+	{
+	  CloseHandle (repl.to_slave); /* not used. */
+	  to = repl.to_slave_nat;
+	}
       else
-	to = repl.to_slave;
+	{
+	  CloseHandle (repl.to_slave_nat); /* not used. */
+	  to = repl.to_slave;
+	}
     }
 
   UINT cp_from = 0, cp_to = 0;
